File: auth-service-go/controllers/auth_controller.go
package controllers

import (
	"auth-service-go/config"
	"auth-service-go/models"
	"auth-service-go/utils"
	"context"
	"time"

	"github.com/gofiber/fiber/v2"
	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/bson/primitive"
)

func Signup(c *fiber.Ctx) error {
	req := c.Locals("signupData").(models.SignupRequest)

	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	collection := config.DB.Collection("users")

	var existingUser models.User
	err := collection.FindOne(ctx, bson.M{"email": req.Email}).Decode(&existingUser)
	if err == nil {
		return utils.ErrorResponse(c, 400, "User already exists with this email", nil)
	}

	hashedPassword, err := utils.HashPassword(req.Password)
	if err != nil {
		return utils.ErrorResponse(c, 500, "Error hashing password", nil)
	}

	user := models.User{
		ID:        primitive.NewObjectID(),
		Email:     req.Email,
		Password:  hashedPassword,
		Name:      req.Name,
		CreatedAt: time.Now(),
		UpdatedAt: time.Now(),
	}

	_, err = collection.InsertOne(ctx, user)
	if err != nil {
		return utils.ErrorResponse(c, 500, "Error creating user", nil)
	}

	token, err := utils.GenerateToken(user.ID.Hex())
	if err != nil {
		return utils.ErrorResponse(c, 500, "Error generating token", nil)
	}

	return utils.SuccessResponse(c, 201, "User registered successfully", models.AuthResponse{
		User:  &user,
		Token: token,
	})
}

func Signin(c *fiber.Ctx) error {
	req := c.Locals("signinData").(models.SigninRequest)

	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	collection := config.DB.Collection("users")

	var user models.User
	err := collection.FindOne(ctx, bson.M{"email": req.Email}).Decode(&user)
	if err != nil {
		return utils.ErrorResponse(c, 401, "Invalid credentials", nil)
	}

	if !utils.ComparePassword(user.Password, req.Password) {
		return utils.ErrorResponse(c, 401, "Invalid credentials", nil)
	}

	token, err := utils.GenerateToken(user.ID.Hex())
	if err != nil {
		return utils.ErrorResponse(c, 500, "Error generating token", nil)
	}

	return utils.SuccessResponse(c, 200, "Login successful", models.AuthResponse{
		User:  &user,
		Token: token,
	})
}

func GetProfile(c *fiber.Ctx) error {
	user := c.Locals("user").(*models.User)

	return utils.SuccessResponse(c, 200, "Profile retrieved successfully", fiber.Map{
		"user": user,
	})
}

File: src/controllers/authController.js
const prisma = require('../config/database');
const { hashPassword, comparePassword } = require('../utils/hashPassword');
const { generateToken } = require('../config/jwt');
const { successResponse, errorResponse } = require('../utils/response');

const signup = async (req, res) => {
  try {
    const { email, password, name } = req.body;

    const existingUser = await prisma.user.findUnique({
      where: { email },
    });

    if (existingUser) {
      return errorResponse(res, 400, 'User already exists with this email');
    }

    const hashedPassword = await hashPassword(password);

    const user = await prisma.user.create({
      data: {
        email,
        password: hashedPassword,
        name,
      },
      select: {
        id: true,
        email: true,
        name: true,
        createdAt: true,
      },
    });

    const token = generateToken(user.id);

    return successResponse(res, 201, 'User registered successfully', {
      user,
      token,
    });
  } catch (error) {
    console.error('Signup error:', error);
    return errorResponse(res, 500, 'Error creating user');
  }
};

const signin = async (req, res) => {
  try {
    const { email, password } = req.body;

    const user = await prisma.user.findUnique({
      where: { email },
    });

    if (!user) {
      return errorResponse(res, 401, 'Invalid credentials');
    }

    const isPasswordValid = await comparePassword(password, user.password);

    if (!isPasswordValid) {
      return errorResponse(res, 401, 'Invalid credentials');
    }

    const token = generateToken(user.id);

    return successResponse(res, 200, 'Login successful', {
      user: {
        id: user.id,
        email: user.email,
        name: user.name,
        createdAt: user.createdAt,
      },
      token,
    });
  } catch (error) {
    console.error('Signin error:', error);
    return errorResponse(res, 500, 'Error during login');
  }
};

const getProfile = async (req, res) => {
  try {
    return successResponse(res, 200, 'Profile retrieved successfully', {
      user: req.user,
    });
  } catch (error) {
    console.error('Get profile error:', error);
    return errorResponse(res, 500, 'Error retrieving profile');
  }
};

module.exports = { signup, signin, getProfile };
